File: ObjectiveC/DFCI-Lib/Carre20Km.m
//
//  Carre20Km.m
//  DFCI-ObjC
//
//  Created by Patrice Trognon on 19/09/2015.
//  Copyright © 2015 Patrice Trognon. All rights reserved.
//

#import "Carre20Km.h"

#import "Globals.h"

#import "LambertII.h"
#import "Coord.h"
#import "Carre2Km.h"

@interface Carre20Km ()
@property(nonatomic,strong)     NSMutableArray          *   matriceCarres2Km;
@end

@implementation Carre20Km

-(instancetype)initWithDFCI:(NSString*)dfci carre100Km:(Carre100Km*)carre100Km {
    self = [super init];
    if ( self ) {
        self.nomDFCIComplet = dfci ;
        self.nomDFCICarre20Km = [dfci substringWithRange:NSMakeRange(0, 4)];
        
        self.carre100Km = carre100Km ;
        
        NSString *dfci20Km = [NSString stringWithFormat:@"%@A0.4",[dfci substringWithRange:NSMakeRange(0, 4)]];
        
        //---
        //   calculer le point SO en lambert
        LambertII *lambert = [self lambertIIFromDFCI:dfci20Km];
        LatLng *latLng = [self latLngFromLambertII:lambert];
        self.coordSO = [[Coord alloc]initWithLatLng:latLng lambertII:lambert];
        
        //---
        //   depuis le point SO calcul des autres points dans un carré de 20 000 m
        {
            LambertII *lambertIISE = [[LambertII alloc]initWithXLambert:lambert.x + 20000
                                                               YLambert:lambert.y];
            LatLng *latLngSE = [self latLngFromLambertII:lambertIISE];
            self.coordSE = [[Coord alloc]initWithLatLng:latLngSE lambertII:lambertIISE];
        }
        
        {
            LambertII *lambertIINE = [[LambertII alloc]initWithXLambert:lambert.x + 20000
                                                               YLambert:lambert.y + 20000];
            LatLng *latLngNE = [self latLngFromLambertII:lambertIINE];
            self.coordNE = [[Coord alloc]initWithLatLng:latLngNE lambertII:lambertIINE];
        }
        
        {
            LambertII *lambertIINO = [[LambertII alloc]initWithXLambert:lambert.x
                                                               YLambert:lambert.y + 20000 ];
            LatLng *latLngNO = [self latLngFromLambertII:lambertIINO];
            self.coordNO = [[Coord alloc]initWithLatLng:latLngNO lambertII:lambertIINO];
        }
        
        {
            LambertII *lambertIICentre = [[LambertII alloc]initWithXLambert:lambert.x + 10000
                                                                   YLambert:lambert.y + 10000];
            LatLng *latLngCentre = [self latLngFromLambertII:lambertIICentre];
            self.coordCentre = [[Coord alloc]initWithLatLng:latLngCentre lambertII:lambertIICentre];
        }
        
    }
    return self;
}

-(instancetype)initWithLatitude:(double)latitude longitude:(double)longitude { // TODO
    self = [super init];
    return self ;
}

-(void)performDetails {
    Carre2Km *carre2Km ;
    
    if ( self.matriceCarres2Km == nil || self.matriceCarres2Km.count != 10*10 ) {
        self.matriceCarres2Km = [NSMutableArray array];
        char LambertX[] = "ABCDEFGHKL";
        char LambertY[] = "0123456789";
        for(int x=0; x<10; x++) {
            for(int y=0; y<10;y++) {
                NSString *dfci = [NSString stringWithFormat:@"%@%c%c.4",[self.nomDFCIComplet substringWithRange:NSMakeRange(0, 4)],LambertX[x],LambertY[y]];
                carre2Km = [[Carre2Km alloc]initWithDFCI:dfci carre20Km:self];
                [self.matriceCarres2Km addObject:carre2Km];
            }
        }
    }
}

-(Carre2Km*)carre2KmWithX:(int)x andY:(int)y {
    Carre2Km *carre2Km ;

    [self performDetails];
    
    carre2Km = self.matriceCarres2Km[(x*10)+y];
    
    return carre2Km ;
}

-(Carre20Km*)northCarre20Km { // TODO
    return nil ;
}

-(Carre20Km*)northEastCarre20Km { // TODO
    return nil ;
}

-(Carre20Km*)eastCarre20Km { // TODO
    return nil ;
}

-(Carre20Km*)southEastCarre20Km { // TODO
    return nil ;
}

-(Carre20Km*)southCarre20Km { // TODO
    return nil ;
}

-(Carre20Km*)southWestCarre20Km { // TODO
    return nil ;
}

-(Carre20Km*)westCarre20Km { // TODO
    return nil ;
}

-(Carre20Km*)northWestCarre20Km { // TODO
    return nil ;
}

-(void)dumpWithIndentLevel:(int)indentLevel {
    CustomLog(@"%@Carre20Km {\n",[self indentWithIndentLevel:indentLevel]);
    CustomLog(@"%@    dfci : %@\n",[self indentWithIndentLevel:indentLevel],self.nomDFCICarre20Km);
    [self dumpCoord:self.coordCentre withPrompt:@"coordCentre" indentLevel:indentLevel+2];
    [self dumpCoord:self.coordSO withPrompt:@"coordSO" indentLevel:indentLevel+2];
    [self dumpCoord:self.coordNO withPrompt:@"coordNO" indentLevel:indentLevel+2];
    [self dumpCoord:self.coordNE withPrompt:@"coordNE" indentLevel:indentLevel+2];
    [self dumpCoord:self.coordSE withPrompt:@"coordSE" indentLevel:indentLevel+2];
    if ( self.matriceCarres2Km != nil && self.matriceCarres2Km.count > 0 ) {
        for(Carre2Km *carre2Km in self.matriceCarres2Km) {
            [carre2Km dumpWithIndentLevel:indentLevel+1];
        }
    }
    CustomLog(@"%@}\n",[self indentWithIndentLevel:indentLevel]);
}

@end

File: C++/DFCI-Lib/Carre20Km.cpp
//
//  Carre20Km.cpp
//  DFCI-Lib-CPP
//
//  Created by Patrice Trognon on 28/10/2015.
//  Copyright © 2015 Patrice Trognon. All rights reserved.
//

#include "Carre20Km.hpp"

#include "Globals.h"

#include "Coord.hpp"
#include "Carre2Km.hpp"

#include "LambertII.hpp"
#include "LatLng.hpp"

#include <iostream>
using namespace std;

Carre20Km::Carre20Km(string dfci, Carre100Km *poCarre100Km_) {
    nomDFCIComplet = dfci ;
    nomDFCICarre20Km = dfci.substr(0,4);
    
    poCarre100Km = poCarre100Km_ ;
    
    string dfci20Km = dfci.substr(0,4) + "A0.4";
    
    //---
    //   calculer le point SO en lambert
    LambertII lambert;
    LatLng latLng;
    
    lambertIIFromDFCI(dfci20Km, & lambert);
    latLngFromLambertII(lambert, & latLng);
    piCoordSO = new Coord(&latLng, &lambert);

    //---
    //   depuis le point SO calcul des autres points dans un carré de 100 000 m
    {
        LambertII lambertIISE(lambert.x + 20000,
                              lambert.y);
        LatLng latLngSE;
        latLngFromLambertII(lambertIISE, & latLngSE);
        piCoordSE = new Coord(&latLngSE, &lambertIISE);
    }
    
    {
        LambertII lambertIINE(lambert.x + 20000,
                              lambert.y + 20000);
        LatLng latLngNE;
        latLngFromLambertII(lambertIINE, & latLngNE);
        piCoordNE = new Coord(&latLngNE, &lambertIINE);
    }
    
    {
        LambertII lambertIINO(lambert.x ,
                              lambert.y + 20000);
        LatLng latLngNO;
        latLngFromLambertII(lambertIINO, & latLngNO);
        piCoordNO = new Coord(&latLngNO, &lambertIINO);
    }
    
    {
        LambertII lambertIICentre(lambert.x + 10000,
                                  lambert.y + 10000);
        LatLng latLngCentre;
        latLngFromLambertII(lambertIICentre, & latLngCentre);
        piCoordCentre = new Coord(&latLngCentre, &lambertIICentre);
    }

    for(int x=0; x<10; x++) {
        for(int y=0; y<10;y++) {
            matriceCarre2Km[x][y] = nullptr ;
        }
    }

}

Carre20Km::~Carre20Km() {
    safeDeletePtr(piCoordSO)
    safeDeletePtr(piCoordSE)
    safeDeletePtr(piCoordNE)
    safeDeletePtr(piCoordNO)
    safeDeletePtr(piCoordCentre)

    if ( matriceCarre2Km[0][0] == nullptr ) {
        for(int x=0; x<10; x++) {
            for(int y=0; y<10;y++) {
                safeDeletePtr(matriceCarre2Km[x][y])
            }
        }
    }
}

void
Carre20Km::performDetails() {
    if ( matriceCarre2Km[0][0] == nullptr ) {
        char LambertX[] = "ABCDEFGHKL";
        char LambertY[] = "0123456789";
        for(int x=0; x<10; x++) {
            for(int y=0; y<10;y++) {
                char dfci[9];
                sprintf(dfci, "%s%c%c.4",nomDFCIComplet.substr(0,4).c_str(),LambertX[x],LambertY[y]);
                matriceCarre2Km[x][y] = new Carre2Km(dfci,this);
            }
        }
    }

}

Carre2Km*
Carre20Km::carre2Km(int x, int y) {
    return matriceCarre2Km[x][y] ;
}

Carre20Km*
Carre20Km::northCarre20Km() {
    return nullptr ;
}

Carre20Km*
Carre20Km::northEastCarre20Km() {
    return nullptr ;
}

Carre20Km*
Carre20Km::eastCarre20Km() {
    return nullptr ;
}

Carre20Km*
Carre20Km::southEastCarre20Km() {
    return nullptr ;
}

Carre20Km*
Carre20Km::southCarre20Km() {
    return nullptr ;
}

Carre20Km*
Carre20Km::southWestCarre20Km() {
    return nullptr ;
}

Carre20Km*
Carre20Km::westCarre20Km() {
    return nullptr ;
}

Carre20Km*
Carre20Km::northWestCarre20Km() {
    return nullptr ;
}

void
Carre20Km::dumpWithIndentLevel(int indentLevel) {
    cout << indentWithIndentLevel(indentLevel) << "Carre20Km {" << endl ;
    cout << indentWithIndentLevel(indentLevel) << "    dfci : " << nomDFCICarre20Km << endl ;
    dumpCoordWithPrompt(*piCoordCentre, "coordCentre", indentLevel+2);
    dumpCoordWithPrompt(*piCoordSO, "coordSO", indentLevel+2);
    dumpCoordWithPrompt(*piCoordNO, "coordNO", indentLevel+2);
    dumpCoordWithPrompt(*piCoordNE, "coordNE", indentLevel+2);
    dumpCoordWithPrompt(*piCoordSE, "coordSE", indentLevel+2);
    if ( matriceCarre2Km[0][0] != nullptr ) {
        for(int x=0; x<10; x++) {
            for(int y=0; y<10;y++) {
                matriceCarre2Km[x][y]->dumpWithIndentLevel(indentLevel+1);
            }
        }
    }
    cout << indentWithIndentLevel(indentLevel) << "}" << endl ;
}
